### PR TITLE
chore: add tests

### LINKED ISSUE #2: Add tests
Some tests I can reference for a custom cache driver: https://github.com/unikent/taggedFileCache/tree/master/tests

### PR DESCRIPTION
**Pull request by poppoerika**
Added tests to test `MomentoStore` and `MomentoTaggedCache`.
Also updated the GitHub workflows to align with the recent update in our PHP SDK to run tests on PHP 8.0.

Closes #2

### CODE REVIEW

**Comment by poppoerika** [comment]
@cprice404 - Could you add `ALPHA_TEST_AUTH_TOKEN` to this repo?

**Comment by cprice404** [comment]
> @cprice404 - Could you add `ALPHA_TEST_AUTH_TOKEN` to this repo?

done

**Comment by poppoerika** [comment]
@pgautier404 
I added three more test cases in`MomentoTaggedCacheTest`. 
These should cover my interpretation of how Tagged Cache should work based on the Laravel documentation on [Tagged Cache](https://laravel.com/docs/9.x/cache#cache-tags).

1. When user tries to get a value tagged with multiple tags, user needs to provide those tags when user wants to get the value in the same order as when they are passed to put the value. In other words, `Cache::tags([$tag1, $tag2])->put($key, $value, $ttl)` cannot be gotten with `Cache::tags([$tag2, $tag1])->get($key)`
2. A value put with tags can only be gotten with tags. In other words, user cannot get the value with `Cache::get($key)` for the value put with `Cache::tags([$tag])->put($key, $value, $ttl)`.
3. A value put with tags can only be gotten with the exact same tags. In other words, `Cache::tags([$tag1, $tag2])->put($key, $value, $ttl)` cannot be gotten with `Cache::tags([$tag1])->get($key)`. 

cc @cprice404

**Comment by poppoerika** [comment]
- Added additional tests to cover tagged cache behavior based on the above.
- Assigned max ttl of a day to `setAdd` operations and refresh to `true` under the hood of tagged cache get.
- Updated the tagged cache get/put operations' logic.

File: tests/MomentoStoreTest.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
+<?php
+//
```
Comment (cprice404) [inline tests/MomentoStoreTest.php]: is this file intended to be included in this PR?

File: src/MomentoTaggedCache.php [hunk 2]
```diff
@@ -6,18 +6,25 @@
...
 
     public function put($key, $value, $ttl = null)
     {
         $tags = $this->tags->getNames();
+        if (!self::validateTags($tags)) {
+            return false;
+        }
         $cacheName = $this->store->getCacheName();
+        $newKey = self::createNewKey($tags, $key);
+        $hashedKey = hash("sha256", $newKey);
```
Comment (cprice404) [inline src/MomentoTaggedCache.php:20]: ❤️

File: src/MomentoTaggedCache.php [hunk 3]
```diff
@@ -6,18 +6,25 @@
...
         foreach ($tags as $tag) {
-            $response = $this->store->setAdd($cacheName, $tag, $key, false, $ttl);
-            if (!$response) {
+            $newKeyResponse = $this->store->setAdd($cacheName, $tag, $newKey, true, self::maxTtl);
+            $hashedKeyResponse = $this->store->setAdd($cacheName, $tag, $hashedKey, true, self::maxTtl);
+            if (!$newKeyResponse || !$hashedKeyResponse) {
```
Comment (cprice404) [inline src/MomentoTaggedCache.php]: I don't think you need to store both of these?  I think one or the other should suffice?

File: src/MomentoTaggedCache.php [hunk 4]
```diff
@@ -27,21 +34,51 @@ public function put($key, $value, $ttl = null)
     public function get($key, $default = null): mixed
     {
         $tags = $this->tags->getNames();
+        if (!self::validateTags($tags)) {
+            return false;
```
Comment (cprice404) [inline src/MomentoTaggedCache.php]: Are we not supposed to throw an exception if the tags are invalid?

File: src/MomentoTaggedCache.php [hunk 5]
```diff
@@ -27,21 +34,51 @@ public function put($key, $value, $ttl = null)
...
+        }
         $cacheName = $this->store->getCacheName();
         $value = null;
+        $newKey = self::createNewKey($tags, $key);
+        $hashedKey = hash("sha256", $newKey);
         foreach ($tags as $tag) {
```
Comment (cprice404) [inline src/MomentoTaggedCache.php]: I don't think you need to do any looping anymore?

File: src/MomentoTaggedCache.php [hunk 6]
```diff
@@ -27,21 +34,51 @@ public function put($key, $value, $ttl = null)
...
                     }
                 }
             }
         }
         return $value;
     }
+
+    private function createNewKey($tags, $key): string
+    {
+        $newKey = "";
```
Comment (cprice404) [inline src/MomentoTaggedCache.php]: I think this whole function could be simplified down to something like:

`join($tags, "-") + "-${key}"`

Also, this is a good example of a function that can be static - it doesn't rely on any member variables of the class.  It's also a great example of a function that you could write some basic unit tests for, and those unit tests wouldn't rely on any other parts of the class or caching logic at all.

File: tests/MomentoTaggedCacheTest.php [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        $key = uniqid();
+        $value = uniqid();
+        $tag1 = uniqid();
+        $tag2 = uniqid();
+        $putResult = Cache::tags([$tag1, $tag2])->put($key, $value, 5);
+        $this->assertTrue($putResult, "True was expected but received ${putResult}");
+        $getResult = Cache::tags([$tag1])->get($key);
+        $this->assertNull($getResult, "null was expected but received ${getResult}");
+    }
+}
```
Comment (cprice404) [inline tests/MomentoTaggedCacheTest.php:121]: I was looking at the laravel cache docs and I notice that there is a `flush` operation for tagged caches.

I think that we actually could implement flush for tagged caches because we have an index of all of the keys.

What is the current behavior if someone calls `flush` since we didn't implement the method?  Does it throw an exception?  Or does it silently execute?  If the latter, a user might be confused and think that we successfully flushed when we actually didn't.

You should add a test here that covers your expected behavior of `flush`.  For this PR, that test can just assert that it throws an error or otherwise alerts the user to a failure.

In a subsequent PR we can actually implement flush.  Don't worry about implementing it in this PR, I'll add a ticket to the board.

File: src/MomentoTaggedCache.php (hunk 4 above)
Comment (pgautier404) [reply src/MomentoTaggedCache.php]: The Laravel docs don't discuss validation of tags (or throwing exceptions) at all. By my reading of the doc, we should handle this as a cache miss and return `null` instead.

File: tests/MomentoTaggedCacheTest.php [hunk 8]
```diff
@@ -0,0 +1,124 @@
+<?php
+
+use Illuminate\Support\Facades\Cache;
+
+class MomentoTaggedCacheTest extends BaseTest
+{
+    /**
+     * @covers \Momento\Cache\MomentoTaggedCache
+     */
```
Comment (pgautier404) [inline tests/MomentoTaggedCacheTest.php]: You can just add this once on the class declaration instead of on each function.

File: tests/MomentoStoreTest.php (hunk 1 above)
Comment (pgautier404) [reply tests/MomentoStoreTest.php]: @poppoerika it looks like you can just uncomment these? They look good and they're all passing for me . . .

File: composer.json [hunk 9]
```diff
@@ -12,14 +12,14 @@
...
+    "orchestra/testbench": "^7.11"
+  },
   "autoload": {
     "psr-4": {
       "Momento\\Cache\\": "src/"
     }
   },
-  "config": {
-    "preferred-install": "dist"
-  },
```
Comment (pgautier404) [inline composer.json:22]: Is there a reason you removed this? In the `composer.json` for the SDK, I added the following:

```
  "config": {
    "platform": {
      "php": "8.0"
    },
```
which tells composer to calculate the dependencies in `composer.json` for PHP 8.0 (our minimum supported version) instead of the version of PHP you happen to be running when you run `composer update`. I'd like us to add that to all of our `composer.json` files so packages don't sneak in that are incompatible for some users.

I'm not too concerned about the `preferred-install` property, and I think "dist" is the default anyhow, but I'd definitely appreciate it if you could add the `platform` property.

File: tests/BaseTest.php [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        parent::setUp();
+        app('cache')->extend('momento', function ($app, $config) {
+            $store = new MomentoStore($config['cache_name'], $config['default_ttl']);
+            return app('cache')->repository($store);
+        });
+        $this->app['config']->set('cache.default', 'momento');
+        $this->app['config']->set('cache.stores.momento',
+            [
+                'driver' => 'momento',
+                'cache_name' => env('MOMENTO_CACHE_NAME'),
```
Comment (pgautier404) [inline tests/BaseTest.php:42]: If this environment variable isn't set, the tests all fail with an error that doesn't suggest the cause of the problem at all. Can you please add a check for that here and throw an exception with a helpful message if the env var isn't set?

File: tests/MomentoStoreTest.php (hunk 1 above)
Comment (poppoerika) [reply tests/MomentoStoreTest.php]: Ah sorry, I was running tests just for TaggedCache and commented MomentoStore tests.
I will uncomment them 🙇‍♀️

File: src/MomentoTaggedCache.php (hunk 4 above)
Comment (poppoerika) [reply src/MomentoTaggedCache.php]: Since `MomentoTaggedCache` extends `TaggedCache`, we want to return `bool` for `get`.
But now I think about it, I should not decide this behavior of returning false in case `$tags` are empty.
`tags` method in `MomentoStore` should be able to throw an exception if users pass empty tags based on the source code (more so `func_get_args()` will thrown an error if an argument is empty):
```php
public function tags($names)
    {
        return new TaggedCache($this, new TagSet($this, is_array($names) ? $names : func_get_args()));
    }
```

File: src/MomentoTaggedCache.php (hunk 5 above)
Comment (poppoerika) [reply src/MomentoTaggedCache.php]: My understanding is this:
In case of `Cache::tags(['people', 'artists'])->put('John', $john, $seconds);`, we need to create two sets. 
One is `people` as a set name and the other is `artists` as a set name.
Each will have a hashed version of `people-artists-John`.
And we store the hashed key with the value `$john`.

I think we need to keep track of all tags since users are able to flush all values associated with each tag with `Cache::tags(['people', 'authors'])->flush();`.
In this case, we need all keys from both tags `people` and `authors` to `forget` values associated with those keys in both sets.

File: tests/MomentoTaggedCacheTest.php (hunk 7 above)
Comment (poppoerika) [reply tests/MomentoTaggedCacheTest.php:121]: I have not added `flush` to `MomentoTaggedCache` so it will probably throw an exception that does not make sense to users. 
I will add the method and throw an `UnknownError` saying that `flush` is not currently supported until we implement it.

Comment (cprice404) [reply tests/MomentoTaggedCacheTest.php:121]: start by writing a test to see what it does, no need to guess

File: composer.json (hunk 9 above)
Comment (poppoerika) [reply composer.json:22]: Yeah, as you mentioned, I noticed that `dist` was the default for `preferred-install` so I removed it.
But yes, I agree that we should be consistent about PHP version. I'll add this configuration to `composer.json`.

File: src/MomentoTaggedCache.php (hunk 4 above)
Comment (poppoerika) [reply src/MomentoTaggedCache.php]: I wrote a test to put with an empty tag and it was actually success though we don't want encourage this behavior.
I agree that we should return `null` in case users pass an empty tag.

Comment (poppoerika) [reply src/MomentoTaggedCache.php]: But in case of put, `false` will be returned.

File: tests/MomentoTaggedCacheTest.php (hunk 7 above)
Comment (poppoerika) [reply tests/MomentoTaggedCacheTest.php:121]: It turns out that if I call `flush` on tagged cache, I get `Momento\Cache\Errors\UnknownError: forever operations is currently not supported.`.
Since `flush` is not implemented in `MomentoTaggedCache`, it goes to `forever` operation which throws the above error.
The below are the stack trace:
```
/Users/erikatharp/laravel-cache/src/MomentoStore.php:94
/Users/erikatharp/laravel-cache/vendor/laravel/framework/src/Illuminate/Cache/TagSet.php:54
/Users/erikatharp/laravel-cache/vendor/laravel/framework/src/Illuminate/Cache/TagSet.php:43
/Users/erikatharp/laravel-cache/vendor/laravel/framework/src/Illuminate/Cache/TaggedCache.php:81
```

Comment (poppoerika) [reply tests/MomentoTaggedCacheTest.php:121]: The error thrown for calling `flush` is confusing, so I will throw an error with this message:
```
 flush operations is currently not supported.
```

Comment (poppoerika) [reply tests/MomentoTaggedCacheTest.php:121]: The error thrown for calling `flush` is confusing, so I will throw an error with this message:
```
 flush operations is currently not supported.
```

File: src/MomentoTaggedCache.php (hunk 5 above)
Comment (pgautier404) [reply src/MomentoTaggedCache.php]: Chris is right. This is a `get`, and we already have all the information we need to retrieve the value from the cache: the `$hashedKey`. This can be simplified down to using `$this->store->get()` to try to get the value from the cache using the hashed key. If we get a value back, we return it. If we don't, we return `$default`. We only ever need to use the sets when handling `put` and `flush`.

File: src/MomentoStore.php [hunk 11]
```diff
@@ -20,22 +20,26 @@ public function __construct(string $cacheName, int $defaultTtl)
         $this->client->createCache($cacheName);
     }
 
-    public function get($key)
+    public function get($key): ?string
```
Comment (pgautier404) [inline src/MomentoStore.php:23]: Thanks for the type hints and doc comments!

File: src/MomentoTaggedCache.php (hunk 5 above)
Comment (poppoerika) [reply src/MomentoTaggedCache.php]: Ah! Somehow I thought it was for `put`. Yes, you're both right about not needing the for loop.

File: tests/BaseTest.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Setup the test environment.
+     */
+    protected function setUp(): void
+    {
+        if (!getenv('MOMENTO_CACHE_NAME')) {
+            throw new UnknownError("An environment variable named MOMENTO_CACHE_NAME must be set.");
+        } else if (!getenv('MOMENTO_AUTH_TOKEN')) {
+            throw new UnknownError("An environment variable named MOMENTO_AUTH_TOKEN must be set.");
```
Comment (pgautier404) [inline tests/BaseTest.php]: Kind of a minor nitpick, but if a user doesn't have either environment variable set, they'll run into two errors before they get straightened out. It might be slightly more helpful to check if either one is missing and, if so, tell the user about both variables in the error message.

File: src/MomentoTaggedCache.php (hunk 4 above)
Comment (pgautier404) [reply src/MomentoTaggedCache.php]: `TaggedCache` doesn't have a `get` method to override. The `get` method here is implementing the definition in `CacheInterface::get`, which has a return value of `mixed`. That makes sense, because if we had to return a `bool` we could never return an actual cache value.

I think this works for now, but the more I think about it the more I think it might make sense to throw an exception here instead. Because the only thing that makes a tag technically invalid is it being empty, and I can imagine how it might be easy to accidentally pass empty tags, it seems like it might be better to alert the user of the error than to return the equivalent of a cache miss. I don't want to hang up the PR for this issue and I think this is good for now, but I'd like to discuss it a little more at some point.

File: src/MomentoTaggedCache.php [hunk 13]
```diff
@@ -27,21 +34,32 @@ public function put($key, $value, $ttl = null)
...
-                    }
-                }
-            }
+        if (!self::validateTags($tags)) {
+            return null;
         }
-        return $value;
+        $newKey = self::createNewKey($tags, $key);
+        $hashedKey = hash("sha256", $newKey);
+        return $this->store->get($hashedKey);
```
Comment (pgautier404) [inline src/MomentoTaggedCache.php]: If the return value here is `null`, we're supposed to return `$default` instead. That variable is `null` by default but might be overridden by the user.

Comment (poppoerika) [reply src/MomentoTaggedCache.php]: Sorry forgot to address that. Thank you for pointing it out!

Comment (poppoerika) [reply src/MomentoTaggedCache.php]: I'll add an additional test to check this behavior as well.

**Review by pgautier404 — DISMISSED** [review]

File: tests/MomentoStoreTest.php [hunk 14]
```diff
@@ -1,75 +1,61 @@
...
+    {
+        $key = uniqid();
+        $value = uniqid();
+        $putResult = Cache::put($key, $value, 5);
+        $this->assertEquals($value, $putResult, "${value} was expected but received ${putResult}");
+        $forgetResult = Cache::forget($key);
+        $this->assertTrue($forgetResult, "True was expected but received ${forgetResult}");
+    }
+
+    public function testTaggedPutGet_HappyPath()
```
Comment (cprice404) [inline tests/MomentoStoreTest.php]: this test might be redundant now that you have a separate test file for tagged cache?

File: tests/UtilTest.php [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * @covers \Momento\Cache\MomentoTaggedCache\
+ */
+class UtilTest extends BaseTest
+{
+    public function testCreateNewKey()
+    {
+        $tags = [uniqid(), uniqid(), uniqid()];
+        $key = uniqid();
+        $expectedNewKey = join("-", $tags) . "-${key}";
```
Comment (cprice404) [inline tests/UtilTest.php]: This test is kind of just testing the implementation, and if there was a bug in how we were doing the join but we happened to be using the same join statement in both places then the test would still pass even if it wasn't doing what we actually expected.

For a test like this it will usually be more useful if you just use actual strings, like:

```php
$tags = ['foo', 'bar'];
$key = ['baz'];
$expectedNewKey = 'foo-bar-baz';
```

That way I can look at the test code and understand exactly what is supposed to be happening without reading the implementation details of the join statement.

**Review by cprice404 — COMMENTED** [review]
just a couple of minor nits about the tests and then we can merge this.

we should have a follow-up to chat about some implications of the TTLs for the tag sets but that's not blocking for this PR.

File: src/MomentoTaggedCache.php [hunk 16]
```diff
@@ -3,21 +3,29 @@
 namespace Momento\Cache;
 
 use Illuminate\Cache\TaggedCache;
+use Momento\Cache\Errors\UnknownError;
 
 class MomentoTaggedCache extends TaggedCache
 {
+    const maxTtl = PHP_INT_MAX;
```
Comment (pgautier404) [inline src/MomentoTaggedCache.php]:
```suggestion
    const MAX_TTL = intdiv(PHP_INT_MAX, 1000));
```

File: src/MomentoTaggedCache.php [hunk 17]
```diff
@@ -3,21 +3,29 @@
...
+            return false;
+        }
         $cacheName = $this->store->getCacheName();
+        $newKey = self::createNewKey($tags, $key);
+        $hashedKey = hash("sha256", $newKey);
         foreach ($tags as $tag) {
-            $response = $this->store->setAdd($cacheName, $tag, $key, false, $ttl);
-            if (!$response) {
+            # Divide PHP_INT_MAX by 1000 so that when mxTtl is converted to milliseconds by ttlToMillis, the return value is still int not float.
+            $hashedKeyResponse = $this->store->setAdd($cacheName, $tag, $hashedKey, true, intdiv(PHP_INT_MAX, 1000));
```
Comment (pgautier404) [inline src/MomentoTaggedCache.php]:
```suggestion
            $hashedKeyResponse = $this->store->setAdd($cacheName, $tag, $hashedKey, true, MAX_TTL));
```

**Comment by poppoerika** [comment]
@pgautier404 
It looks like PHP doesn't let me use an expression `intdiv` as a constant or class field property. 
So I defined `$MAX_TTL` variable inside of the `put` method for `MomentoTaggedCache`.
I'm not sure if there is a different way to workaround this restriction, so your input would be appreciated 🙇‍♀️

File: src/MomentoTaggedCache.php [hunk 18]
```diff
@@ -19,7 +18,8 @@ public function put($key, $value, $ttl = null): bool
         $newKey = self::createNewKey($tags, $key);
         $hashedKey = hash("sha256", $newKey);
         foreach ($tags as $tag) {
-            $hashedKeyResponse = $this->store->setAdd($cacheName, $tag, $hashedKey, true, self::maxTtl);
+            # Divide PHP_INT_MAX by 1000 so that when mxTtl is converted to milliseconds by ttlToMillis, the return value is still int not float.
```
Comment (cprice404) [inline src/MomentoTaggedCache.php]: this comment seems like it goes with the variable assignment up above, rather than here?

**Comment by pgautier404** [comment]
> @pgautier404 It looks like PHP doesn't let me use an expression `intdiv` as a constant or class field property. So I defined `$MAX_TTL` variable inside of the `put` method for `MomentoTaggedCache`. I'm not sure if there is a different way to workaround this restriction, so your input would be appreciated 🙇‍♀️

Sure, that works for me.

**Comment by pgautier404** [comment]
I'm good with this if @cprice404 is!

**Review by cprice404 — APPROVED** [review]
